### PR TITLE
Use plan graph for importing resources

### LINKED ISSUE #27934: Provider configuration cannot depend on data sources during import
As documented in [Import: Provider Configuration](https://www.terraform.io/docs/cli/commands/import.html#provider-configuration), a provider used during import cannot depend on data sources for its configuration*. 

As of v0.14, data sources can now be handled entirely during planning, which in principle works similarly to the import process. This means that it should be possible to expand and evaluate required data sources during import as well, as long as they themselves do not depend on any resource data which is not contained within the state.

---

*This shortcoming is no longer entirely true, a data source can be read from the existing state. The workaround for this case is to add the data source and refresh it immediately before running import. A state is required to refresh however, so if there has yet to be any created, applying a single `null_resource` will allow the `refresh` to complete, reading the data source into the state. 

References:
This was mentioned in the comments of #13018, but the original cases presented there are no longer relevant in current releases.

### LINKED ISSUE #30706: Importing a resource doesn't work if the module is calling other modules via for_each
### Terraform Version

```
1.1.7
```

### Terraform Configuration Files
https://github.com/applike-ss/tf-import-issue

### Debug Output
https://gist.github.com/applike-ss/060f787190d29b707c06a0efccecdaa4

### Expected Behavior
resource should be importable, because the for_each does not matter for the resource

### Actual Behavior
resource was prepared to import, but import failed with `Invalid for_each argument` and `data.null_data_source.f will be known only after apply`

### Steps to Reproduce
* check out repo
* have an aws account (or exchange the resource against anything else that can be imported that you have an account for)
* `terraform init`
* `terraform import 'module.m[0].aws_sns_topic.t' arn:aws:sns:eu-central-1:123456789012:my-topic`

### PR DESCRIPTION
**Pull request by jbardin**
The first thing done here is that we combine the functionality of all the plan-time graph builders, namely `plan`, `destroy`, and `validate` into a single `GraphBuilderPlan` for a single entrypoint. This is mostly a continuation of the process started in #31163 to consolidate graph building behaviors.

Once we have a single plan graph, we then add the ability to deal with imports, which turns out to be fairly trivial. Rather than using a "transformer" to modify the graph and add individual import nodes, we add the import addresses to the abstract resource nodes, and let them create the correct node during resource expansion. This ensures all instances are accounted for, even if they are not currently planning or refreshing anything.

For now we skip planning and refreshing other instances from config during an import operation. Instance planning will probably need to be added for more complex configurations where the planned values need to be evaluated, but is disabled to try and stay as close the the current behavior as possible for testing (the unit test failures from enabling that are likely all just incomplete test fixtures). Refresh cannot be added until the import process can be fully "planned", since the current result of planning uses the resulting state directly.

Further integrating this into `Context.Plan`, and making import act more like a plan option should probably be done in concert with adding import actions to the plan itself, so that we can have the ability to store, compare, and display the imported state rather than blindly overwriting the existing state.

Due to the necessity of having existing configuration for more complete planning of imports, the `-allow-missing-config` option is being removed from the import command. This possibility was allowed for in the *v1.0 Compatibility Promises* under [commands that might change](https://www.terraform.io/language/v1-compatibility-promises#commands-that-might-change). 

Fixes #30706
Fixes #27934
May apply to #30746 as well, though we don't have a reproduction for that yet.

### CODE REVIEW

**Comment by jbardin** [comment]
The test failures are related to the old `-allow-missing-config` option, which obviously doesn't work well with the planning process which is based on configuration. Will follow up once we evaluate how that can be handled in a compatible manner.

**Review by alisdair — APPROVED** [review]
Neat! I smoke tested this briefly locally and all seemed well.

A couple of minor thoughts inline from reading the diff.

File: internal/terraform/graph_builder_plan.go [hunk 1]
```diff
@@ -210,5 +227,61 @@ func (b *PlanGraphBuilder) init() {
...
+			// For now we always skip planning changes for import, since we are
+			// not going to combine importing with other changes. This is
+			// temporary to try and maintain existing import behaviors, but
+			// planning will need to be allowed for more complex configurations.
+			skipPlanChanges: true,
+
+			// We also skip refresh for now, since the plan output is written
+			// as the new state, and users are not expecting the import process
+			// to update any other instances in state.
+			skipRefresh: true,
```
[the remark this answers is not in the document]
Comment (jbardin) [reply internal/terraform/graph_builder_plan.go:284]: Thanks, that's just an oversight, and should be synced up in some way 👍

**Comment by jbardin** [comment]
I'm going to merge this as-is for now so that there's more time to try and shake out any unexpected issues with the already fiddly import process. I will still follow up with the possibility of re-instating the `-allow-missing-config` for broader compatibility with existing workflows.

**Pull request merged**